### PR TITLE
Migrate cancel

### PR DESCRIPTION
**Pull request by Howard-Chang**
Add new hmp command "cuju-migrate-cancel".

### CODE REVIEW

File: linux-headers/linux/kvm.h [hunk 1]
```diff
@@ -1415,6 +1416,7 @@ struct kvmft_set_master_slave_sockets {
 };
 #define KVMFT_SET_MASTER_SLAVE_SOCKETS    _IOW(KVMIO, 0xcf, struct kvmft_set_master_slave_sockets)
 
+#define KVM_SHM_CANCEL_TIMER       _IO(KVMIO, 0xd2)
```
Comment (PJTPJT) [inline linux-headers/linux/kvm.h]: It same as #define KVM_RESET_PUT_OFF _IOW(KVMIO, 0xd2, int)
please change another numver

File: migration/cuju-ft-trans-file.c [hunk 2]
```diff
@@ -108,6 +117,13 @@ void cuju_ft_trans_flush_buf_desc(void *opaque)
                 //break;
                 continue;
             } else if (ret <= 0) {
+                if(s->check)
+                {
+                    s->check = false;
+                    //printf("cuju_ft_trans_flush_buf_desc!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n");
```
Comment (PJTPJT) [inline migration/cuju-ft-trans-file.c]: Please fix this debug message

File: migration/cuju-ft-trans-file.c [hunk 3]
```diff
@@ -574,7 +600,13 @@ static int cuju_ft_trans_try_load(CujuQEMUFileFtTrans *s)
 #endif
 
     while (s->ft_serial == ft_serial && cuju_ft_trans_load_ready(s)) {
-        ret = cuju_ft_trans_send_header(s, CUJU_QEMU_VM_TRANSACTION_ACK1, 0);
+        ret = cuju_ft_trans_send_header(s,s->check<<CUJU_FT_ALIVE_HEADER|CUJU_QEMU_VM_TRANSACTION_ACK1, 0);
```
Comment (PJTPJT) [inline migration/cuju-ft-trans-file.c]: Please add some description for this part
(or maybe copy CUJU_FT_ALIVE_HEADER define comment)

File: migration/migration.c [hunk 4]
```diff
@@ -2117,7 +2158,7 @@ bool migrate_cuju_enabled(void)
     return s->enabled_capabilities[MIGRATION_CAPABILITY_CUJU_FT];
 }
 
-static void migrate_fd_get_notify(void *opaque)
+/*static void migrate_fd_get_notify(void *opaque)
```
Comment (PJTPJT) [inline migration/migration.c]: Please do not comment out original migrate_fd_get_notify and create a new one
Direct modify this function

File: migration/cuju-kvm-share-mem.c [hunk 5]
```diff
@@ -1098,8 +1129,17 @@ static void* trans_ram_conn_thread_func(void *opaque)
         s->ram_len += ret;
 
         ret = kvm_start_kernel_transfer(s->cur_off, s->ram_fds[d->index], d->index, ft_ram_conn_count);
-
-        assert(ret >= 0);
+        if(ret<0)
+        {
+            //printf("ret<0    %d!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n",ret);
```
Comment (PJTPJT) [inline migration/cuju-kvm-share-mem.c]: Please refine this description

File: migration/event-tap.c [hunk 6]
```diff
@@ -612,9 +612,9 @@ int event_tap_flush(void* cb, void *opaque)
         ret = event_tap_flush_one();
     } while (ret == 0);
 
-    if (pending_bdrv_request == 0) {
```
Comment (PJTPJT) [inline migration/event-tap.c]: Why comment out this part?

File: migration/migration.c [hunk 7]
```diff
@@ -1403,7 +1417,8 @@ MigrationState *migrate_init(const MigrationParams *params)
 
     s->total_time = qemu_clock_get_ms(QEMU_CLOCK_REALTIME);
     s2->total_time = qemu_clock_get_ms(QEMU_CLOCK_REALTIME);
-
+    s->ft_state = CUJU_FT_OFF;
```
Comment (PJTPJT) [inline migration/migration.c]: ?

File: linux-headers/linux/kvm.h (hunk 1 above)
Comment (Howard-Chang) [reply linux-headers/linux/kvm.h]: I have modified the magic number to 0xdd, and there's no conflict.

File: migration/cuju-ft-trans-file.c (hunk 2 above)
Comment (Howard-Chang) [reply migration/cuju-ft-trans-file.c]: fix it to printf("primary has sent checkalive header and %s ret <= 0 line:%d\n", __func__, __LINE__);

File: migration/cuju-ft-trans-file.c (hunk 3 above)
Comment (Howard-Chang) [reply migration/cuju-ft-trans-file.c]: add some description:
/*
            if backup receive checkalive header then s->check = 1 and the ACK1 header would set leftmost bit to be 1.   
            so s->check<<CUJU_FT_ALIVE_HEADER equal to 1<<15.
*/

File: migration/migration.c (hunk 4 above)
Comment (Howard-Chang) [reply migration/migration.c]: add a new function which calls "cuju_migrate_fd_get_notify" and comment out the original "migrate_fd_get_notify" function

File: migration/cuju-kvm-share-mem.c (hunk 5 above)
Comment (Howard-Chang) [reply migration/cuju-kvm-share-mem.c]: fix it to printf("%s ret<0 \n", __func__);

File: migration/event-tap.c (hunk 6 above)
Comment (Howard-Chang) [reply migration/event-tap.c]: if no comment out this part, it would occur segmentation fault.

File: migration/migration.c (hunk 7 above)
Comment (Howard-Chang) [reply migration/migration.c]: if no initialize s->ft_state to CUJU_FT_OFF, after cuju-migrate-cancel the state would become CUJU_FT_TRANSACTION_FLUSH_OUTPUT and occur error:
qemu-system-x86_64: migration/migration.c:133: migrate_assert_ft_state_change: 
Assertion `t == CUJU_FT_TRANSACTION_FLUSH_OUTPUT' failed.

File: migration/migration.c (hunk 4 above)
Comment (PJTPJT) [reply migration/migration.c]: original migrate_fd_get_notify？

Comment (Howard-Chang) [reply migration/migration.c]: because original migrate_fd_get_notify not used if I don't comment out it would get error:
not used but declared.

**Pull request merged**